### ISSUE TITLE
Race condition with remove and rename

### ISSUE THREAD

**Issue by Nikratio** [syncthing/syncthing#8036]
I recently encountered the following in a synchronized folder:

```
$ rm foo
$ sleep 18 # (estimated, not sure how long I waited)
$ mv bar foo
$ # do nothing for a day
$ ls foo
No such file or directory
```

Upon further investigation, I found the contents of `bar` as `foo` in the `.stversions` folder.

I assume this means that the `foo` file was deleted by Syncthing response to it being deleted on another device.

However, I did not delete (or even touch) it on any other device.

Is it possible that the remove and rename operations somehow raced, resulting in the other device first executing the `mv` and then the `rm'?

I have not been able to reproduce this problem, and I understand that I am probably providing insufficient information to investigate this further. However, given the potential severity of this issue I wanted to give a heads-up, and enable others who may have experienced something similar but blamed a human error to chime in.

I am using Syncthing  v1.12.1-ds1 on Linux on the device where I observed this. Other devices run different syncthing versions v1.18.3 and v1.15.1 on Android.

**Comment by acolomb** [comment]
In fact I have experienced the exact same problem just a few days ago.  Was very glad to have one device where the lost file could be recovered from `.stversions`.

**Comment by AudriusButkevicius** [comment]
I am pretty sure you will not be able to reproduce this without android :/

**Comment by acolomb** [comment]
In my case, Android was definitely involved with one or two connected devices.  I noticed (and recovered from) it on a Linux machine, but at least one of the Androids was connected at the time.

**Comment by tomasz1986** [comment]
> I am pretty sure you will not be able to reproduce this without android :/

Could this be related to the infamous Android moving timestamps? If yes, could https://github.com/syncthing/syncthing/commit/8d8f331a4a9b8458991db6e04ca7a7fb65d51235 help?

The fix has been present since v1.18.3, but the OP is using an older version on his Android device.

**Comment by imsodin** [comment]
Unless anyone can repro without android, this is most likely https://github.com/syncthing/syncthing-android/issues/1801 (which likely will be fixed soon). Thus closing here.
